### DISCUSSION TITLE
2个 C API的问题

### DISCUSSION THREAD

**Discussion by stopandjoin** [Tencent/tquic#192] [Q&A (问题和回答)] [answered]
改了一下[tquic-example-c](https://github.com/tquic-group/tquic-example-c)  中的代码，验证多流的收发，遇到两个问题：
1， 发现client端使用 quic_stream_new 创建新流时， stream_id 只能是偶数， 如果是奇数就会返回 -5 错误码。
2， client 在client_on_conn_established 回调中，使用quic_stream_new 创建三个流， 然后在client_on_stream_created 回调中，通过quic_stream_write 分别对各个流写入测试数据。   在server端 server_on_stream_readable回调中，读完流后，立即通过quic_stream_write向对应的流写入测试数据。 发现如下现象: 
server端可以正常收到 client端三个流的数据， 在回写数据时， 前两个流quic_stream_write的返回都是 -5 ，最后一个流可以成功发送。

请大家协助看看，谢谢。

**Comment by glime-ai** [comment]
针对您遇到的两个问题，这里提供一些可能的解释和建议：

1. **关于`stream_id`只能是偶数的问题**：在QUIC协议中，客户端和服务器端使用的流ID有不同的奇偶性要求。具体来说，客户端发起的流使用偶数ID，而服务器发起的流使用奇数ID。这是QUIC协议设计的一部分，用于区分流的发起方。因此，当您在客户端尝试使用奇数`stream_id`创建流时，会遇到错误，这是符合QUIC协议规范的。您可以查阅QUIC协议的相关文档，了解更多关于流ID分配的规则。

2. **关于`quic_stream_write`返回-5的问题**：错误码-5通常表示某种类型的错误，但具体含义可能依赖于TQUIC库的实现细节。一种可能的原因是流状态不允许写入操作。例如，如果流已经被关闭或重置，尝试在该流上执行写入操作可能会失败。另一种可能性是流的流控制限制，即流的发送窗口已满，暂时无法发送更多数据。

针对这个问题，建议的解决步骤包括：
- 确认流的状态，确保在尝试写入数据之前流是处于开放状态的。
- 检查流控制信息，确保流的发送窗口允许发送更多数据。
- 查看TQUIC库的文档或源代码，了解错误码-5的具体含义，以及如何查询流的状态和流控制信息。

由于这些解释和建议是基于QUIC协议的通用知识，而不是特定于TQUIC库的实现细节，建议您查阅[TQUIC的官方文档](https://github.com/tquic-group/tquic-example-c)和相关API参考，以获取更准确的信息和解决方案。如果问题仍然存在，考虑在TQUIC的GitHub仓库中提出issue，以获得开发者的直接支持。

**Comment by iyangsj** [comment, marked as answer]
1. 客户端发起的流ID[需使用偶数](https://datatracker.ietf.org/doc/html/rfc9000#name-stream-types-and-identifier)。其中双向流使用0/4/8/..., 单向流使用2/6/10/...
<img width="697" alt="image" src="https://github.com/Tencent/tquic/assets/780807/1ea9556c-8e42-44d8-a2ad-9a6cd74a0b8e">

2. 服务端在向流写入时，取决于写入数据大小，可能仅写入成功部分数据。
剩余数据应在on_stream_writable回调中继续写入。

返回值为-5，代表StreamStateError，可能由于客户端创建的并非双向流
https://github.com/Tencent/tquic/blob/develop/src/error.rs#L192C20-L192C36

**Reply by iyangsj** [reply]
后续版本会提供改进的接口：
https://github.com/Tencent/tquic/issues/193
